### PR TITLE
unicafe project

### PR DESCRIPTION
**Pull request by chrisOc93**
URL for exercise - https://fullstackopen.com/en/part1/a_more_complex_state_debugging_react_apps#exercises-1-6-1-14

Working with props and state

### CODE REVIEW

**Comment by chrisOc93** [comment]
Ill add some changes so please leave review for now

File: e1.6-1.14/src/App.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  return (
+    <div>
+      <Header/>
+      <Buttons/>  
+     
+    </div>
+  )
+}
+
+export default App
```
Comment (john-behan-tixserve) [inline e1.6-1.14/src/App.js]: @chrisOc93 if you use VS Code set this - https://stackoverflow.com/a/54270768 to automatically add a new line to the end when saving a file

**Comment by chrisOc93** [comment]
@john-behan-tixserve Review can begin now for the first two exercises, I'm still working on the anecdotes exercise so that can be ignored for now. Thanks

**Review by hylee-tixserve — APPROVED** [review]
Hi Chris,

I'm going to focus on the unicafe assignment as I know you're still working on the anecdotes. It's a pretty straightforward exercise and you've done a good job with some nice clear code.

A few more general things following on from the conversation we had earlier today:

- Pay attention to how you're indenting the code — as well as making everything look neater, it really helps you (and others) read your code as you can see straight away where functions start and end, what is nested inside something else, and so on. I'd go as far as to say you make assumptions about the hierarchy based on the indentations, and you have to concentrate quite hard to understand eg that a line of code that looks nested inside a function based on its indentation actually isn't. Like seeing the word 'blue' in red ink.

- This also goes for spaces between lines of code. Use line breaks to separate out chunks of code and make it easier for you to read, but I wouldn't add breaks in the middle of a function, and certainly not several lines, as it all starts to look disconnected. 

- I'd also recommend putting semicolons at the end of code even where it's optional. Like the previous point, I think it helps you scan your code quickly — but I'll also be clear that this may be my personal opinion because I'm used to seeing semicolons there, and many excellent JS developers don't see the need for them. In any case, the linters (tools imported into VSCode — like code spellcheck) will add them in for you if you don't. Whichever you choose to do, I'd suggest you do it consistently — in the Buttons component, for example, you have semicolons at the end of the variables on lines 28 to 30 but none elsewhere. 

- Finally, it's a good idea to pay attention to spelling in variable names. It looks a like a small thing — like 'goodPrecent' rather than 'goodPercent' (l.28-30, Buttons component) — but a misspelled or poorly-named variable can obscure what you're trying to do in the code. It's like the old Ronseal advert — it does what it says on the tin — which is great, except when you don't understand what the label means.

Hyun

File: anecdotes/src/App.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    'Premature optimization is the root of all evil.',
+    'Debugging is twice as hard as writing the code in the first place. Therefore, if you write the code as cleverly as possible, you are, by definition, not smart enough to debug it.',
+    'Programming without an extremely heavy use of console.log is same as if a doctor would refuse to use x-rays or blood tests when diagnosing patients'
+  ]
+   
+  const [selected, setSelected] = useState(0)
+  function randomItem () {
+    const randomNumber = Math.floor(Math.random() * anecdotes.length);
+      setSelected(randomNumber);
+   console.log(selected)
```
Comment (john-behan-tixserve) [inline anecdotes/src/App.js:19]: This won't log the correct number consistently as the `useState` function won't have updated the `selected` value.

This is the section in the docs that covers this - https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous

**Review by john-behan-tixserve — APPROVED** [review]
Hi Chris,

Looks good. There are a couple of things you can do to improve it
 - it looks like you have build the second exercise on top of the branch for the first exercise. This won't cause any problems now but in future it would be better to create a new branch for each exercise.
 - We covered the state updates during the week but I've added a link to the React docs about this
 - For the "Vibe" exercise there is an opportunity to break the "current vibe" parts into a single component that you can import and reuse. SO:
 ```
<tr >
        <td>Positive: {good}</td>
        <td>out of {total} -</td>
        <td>{goodPrecent}%</td>
  </tr>
```
Would become something like:
```
<Vibe title='Positive' total={total} setTotal={setTotal} />
```
I haven't tested this but the Vibe component would then handle it's own value and percent.

We can have a look at this when I get back or maybe John or Hyun can help during the week with this idea.

I also agree with Hyun on the styling side. I use semicolons but in most cases it's optional.

Overall I think this is good and you've learned a lot in the past few weeks.

I'll leave this open for further review as the anecdotes exercise is still in progress. But once that's done I think this can be merged.

John

**Pull request merged**